rust/lance/src/index/vector/ivf/v2.rs

```diff
@@ -0,0 +1,508 @@
+
+        //     }
+        //     arrow::datatypes::DataType::Float32 => {
+        //         part_index.search(query.key, query.k, (&query).into(), &storage, pre_filter)
+        //     }
+        //     arrow::datatypes::DataType::Float64 => {
+        //         part_index.search(query.key, query.k, (&query).into(), &storage, pre_filter)
+        //     }
+        //     _ => Err(Error::Index {
+        //         message: "unsupported data type".to_string(),
+        //         location: Location::new(file!(), line!(), column!()),
+        //     }),
+    }
+
+    /// preprocess the query vector given the partition id.
+    ///
+    /// Internal API with no stability guarantees.
+    pub fn preprocess_query(&self, partition_id: usize, query: &Query) -> Result<Query> {
+        if I::use_residual() {
+            let partition_centroids = self.ivf.centroids.value(partition_id);
+            let residual_key = sub(&query.key, &partition_centroids)?;
+            let mut part_query = query.clone();
+            part_query.key = residual_key;
+            Ok(part_query)
+        } else {
+            Ok(query.clone())
+        }
+    }
+
+    pub fn find_partitions(&self, query: &Query) -> Result<UInt32Array> {
+        let dt = if self.distance_type == DistanceType::Cosine {
+            DistanceType::L2
+        } else {
+            self.distance_type
+        };
+
+        self.ivf.find_partitions(&query.key, query.nprobes, dt)
+    }
+}
+
+#[async_trait]
+impl<I: IvfSubIndex + 'static, Q: Quantization + 'static> Index for IVFIndex<I, Q> {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_index(self: Arc<Self>) -> Arc<dyn Index> {
+        self
+    }
+
+    fn index_type(&self) -> IndexType {
+        IndexType::Vector
+    }
+
+    fn statistics(&self) -> Result<serde_json::Value> {
+        let partitions_statistics = self
+            .ivf
+            .lengths
+            .iter()
+            .map(|&len| IvfIndexPartitionStatistics { size: len })
+            .collect::<Vec<_>>();
+
+        let centroid_vecs = centroids_to_vectors(&self.ivf.centroids)?;
+
+        Ok(serde_json::to_value(IvfIndexStatistics {
+            index_type: "IVF".to_string(),
+            uuid: self.uuid.clone(),
+            uri: self.uuid.clone(),
+            metric_type: self.distance_type.to_string(),
+            num_partitions: self.ivf.num_partitions(),
+            sub_index: Default::default(),
+            partitions: partitions_statistics,
+            centroids: centroid_vecs,
+        })?)
+    }
+
+    async fn calculate_included_frags(&self) -> Result<RoaringBitmap> {
+        todo!()
+        // let mut frag_ids = RoaringBitmap::default();
+        // let part_ids = 0..self.ivf.num_partitions();
+        // for part_id in part_ids {
+        //     let part = self.load_partition(part_id, false).await?;
+        //     frag_ids |= part.calculate_included_frags().await?;
+        // }
+        // Ok(frag_ids)
+    }
```

Comment:
You can probably (intentionally) skip this with a message (maybe use `unimplemented` instead of `todo` so its more obvious).  This method was only needed for migrating older manifests that did not correctly track the fragment bitmap.  New indices should not have to worry about this.